File: trupti/elixir/wordle.exs
defmodule Wordle do
  Code.require_file("wordle_httpoison.exs")

  def run do
    all_words = load_words("5words.txt")

    with {:ok, cookie, user_id} <- API.register(),
         {:ok, cookie, game_id} <- API.create(cookie, user_id) do
      game_loop(all_words, cookie, game_id, 1)
    else
      {:error, reason} -> IO.puts("Setup failed: #{reason}")
    end
  end

  def load_words(filepath) do
    filepath
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(&(&1 |> String.trim() |> String.upcase()))
  end

  defp game_loop(_words, _cookie, _game_id, attempt) when attempt > 5,
    do: IO.puts("Bot couldn't solve in 5 attempts.")

  defp game_loop(words, cookie, game_id, attempt) do
    IO.puts("\n--- Attempt #{attempt} ---")
    # IO.puts("Possible words remaining: #{length(words)}")

    if Enum.empty?(words) do
      IO.puts("❌ No possible words left! Check the logic or wordlist.")
      :ok
    else
      guess = Enum.random(words)
      case API.guess(cookie, game_id, String.downcase(guess)) do
        # This is the winning case. We now capture the `data` from the API.
        {:ok, _feedback, true, data, _new_cookie} ->
          # The `data` map contains the API response. We extract the "answer".
          answer = Map.get(data, "answer", "(Answer not found in API response)")
          IO.puts("🎉 Solved! The word was: #{String.upcase(answer)}")

        {:ok, feedback, false, _data, new_cookie} ->
          new_words = dropImpossibles(words, guess, feedback)
          # IO.puts("After filtering, #{length(new_words)} words remain.")
          game_loop(new_words, new_cookie, game_id, attempt + 1)

        {:error, reason} ->
          IO.puts("Guess failed: #{reason}")
      end
    end
  end

  def dropImpossibles(possible_words, guess, feedback) do
    guess_chars = String.to_charlist(guess)
    feedback_chars = String.to_charlist(feedback)

    Enum.filter(possible_words, fn possible_word ->
      rules = Enum.zip(guess_chars, feedback_chars) |> Enum.with_index(0)

      Enum.all?(rules, fn{{guess_char, feedback_char} , index}->
        case feedback_char do
          ?R ->
            not String.contains?(possible_word, <<guess_char>>)
          ?Y ->
            String.contains?(possible_word, <<guess_char>>) and String.at(possible_word,index)!=<<guess_char>>
          ?G ->
            String.at(possible_word, index) == <<guess_char>>
          _ ->
            true
          end
      end)
    end)
  end
end

Wordle.run()

File: trupti/python/wordle.py
import random
import requests
URL = "https://wordle.we4shakthi.in/game/"
REGISTER_URL = URL +  "register"
CREATE_URL = URL + "create"
GUESS_URL = URL + "guess"

def load_words(filepath: str) -> list[str]:
    with open(filepath, 'r') as file:
        words = [line.strip().upper() for line in file]
    return words

def dropImpossibles(guess_word, feedback, remaining_words) -> list[str]:
    greens = {i: guess_word[i] for i, f in enumerate(feedback) if f == 'G'}
    yellows = {guess_word[i] for i, f in enumerate(feedback) if f == 'Y'}
    reds = {guess_word[i] for i, f in enumerate(feedback) if f == 'R'}
    possible_words = [
        word for word in remaining_words
        if all(word[i] == char for i, char in greens.items()) and
           all(char in word for char in yellows) and
           all(char not in word for char in reds)
    ]
    return possible_words


def register_and_create_game(session: requests.Session) -> str | None:

    try:
        register_data = {"mode": "wordle", "name": "GeminiBot"}
        reg_response = session.post(REGISTER_URL, json=register_data, timeout=10)
        reg_response.raise_for_status()
        game_id = str(reg_response.json().get("id"))

        create_data = {"id": game_id, "overwrite": True}
        create_response = session.post(CREATE_URL, json=create_data, timeout=10)
        create_response.raise_for_status()
        
        # print(f"Game created with ID: {game_id}")
        return game_id
    except requests.exceptions.RequestException as e:
        print(f"Error setting up game: {e}")
        return None

def make_guess(session: requests.Session, game_id: str, guess_word: str) -> str | None:
    try:
        guess_data = {"guess": guess_word, "id": game_id}
        response = session.post(GUESS_URL, json=guess_data, timeout=10)
        response.raise_for_status()
        guess_json = response.json()
        feedback_list = guess_json.get("feedback", [])
        return "".join(feedback_list)
    except requests.exceptions.RequestException as e:
        print(f"Error making guess: {e}")
        return None
    
def main():
    all_words = load_words("5words.txt")
    remaining_words = all_words[:]

    with requests.Session() as session:
        game_id = register_and_create_game(session)
        if not game_id:
            return
        print()
        for attempt in range(1, 6):
            if not remaining_words:
                print("No possible words left! The bot is stumped.")
                break
            guess_word = random.choice(remaining_words)
            feedback = make_guess(session, game_id, guess_word)
            if not feedback:
                print("Could not retrieve feedback. Exiting.")
                break
            print(f"-- Attempt {attempt} --")
            print(f"Guess    : {guess_word}")
            print(f"feedback : {feedback}")
            print()

            if feedback == "GGGGG":
                print("Yay! The bot solved the Wordle! 🎉")
                break
            
            remaining_words = dropImpossibles(guess_word, feedback, remaining_words)
        else:
            print("Oops! The bot couldn't solve the Wordle in 6 attempts. 😢")

if __name__ == "__main__":
    main()